### PR TITLE
ci: build example-test docs once and shard the test jobs

### PR DESCRIPTION
**Pull request by adrianschmidt**
> **Stacked on #4102.** Base is `test/example-axe-tests`; review #4100 → #4101 → #4102 first. I'll retarget to `main` as the stack merges.

## What

Optimizes the example-test CI jobs introduced in #4101/#4102:

- **Build the docs once.** A new `Build Example Docs` job runs `npm run docs:build` and uploads `www/` as an artifact (1-day retention).
- **Reuse it everywhere.** The test jobs now `needs:` that job and **download** the artifact instead of each rebuilding the docs (~3 min apiece).
- **Shard the test jobs across parallel runners** — **runtime ×2**, **accessibility ×4** (`--shard=$k/N`).

## Why

Each example-test job previously rebuilt the docs independently; now the ~3-min build happens once and the test jobs just download it.

Sharding matters because GitHub's standard runners are 2-core and the suite runs at `workers: '50%'` (≈1 worker), so within-job parallelism is limited — splitting across jobs is what actually cuts wall-clock. Measured on the prior runs: the accessibility job was ~9m06s and the runtime job ~5m14s. Sharding turns those into short parallel jobs.

## Notes

- No test logic changes — workflow only. The suites, baseline, and scripts are unchanged.
- All actions are SHA-pinned with version comments (`actions/cache` → v5.0.5, `upload-artifact` → v7.0.1, `download-artifact` → v8.0.1), matching the pins already used elsewhere in the repo.
- **Where the win comes from.** Most of it is the **accessibility** suite (~9m06s → ~4 min) plus building the docs **once** instead of in both jobs (dropping a duplicated ~3-min build). The **runtime** split is marginal on its own (~5m14s → ~4m05s); its value is keeping runtime level with the four accessibility shards rather than becoming the long pole.
- **Cost.** Total runner-minutes rise: per-job setup (checkout + `npm ci` + cache restore) is now spread over 7 jobs (1 build + 2 runtime + 4 accessibility) instead of 2. But that 7× setup buys the **combined** ~10 min of wall-clock reduction across both suites — it is not the price of the runtime minute alone. Deliberate: PR-feedback latency matters more here than runner-minutes.
- Measured shards are balanced (build ~1.2 min; runtime 2.6–2.8 min; accessibility 2.3–2.9 min), so no shard is the long pole.
- Local check: `--shard=1/4` runs exactly 110 of the 440 examples and passes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Chores**
  * Optimized continuous integration workflow with enhanced artifact caching and parallel test execution to improve build feedback speed and reliability.

### CODE REVIEW

**Comment by adrianschmidt** [comment]
# Consolidated PR Review (7-agent)

## PR Summary

PR #4103 is a workflow-only change to `.github/workflows/pr-checks.yml` (+61/−25, 1 file). It adds a `build-example-docs` job that builds the docs once and shares `www/` as a 1-day artifact, converts the `runtime-tests` (×2) and `accessibility-tests` (×4) jobs into sharded matrices that download the artifact instead of rebuilding, and SHA-pins `actions/cache` plus adds least-privilege `permissions: contents: read` to the new/changed jobs. Stacked on #4102; base is `test/example-axe-tests`.

## Merge Readiness -- READY TO MERGE ✅

This PR is strictly better than its target branch: same test coverage provably preserved across shards, byte-identical docs build, tighter action pinning and permissions, and a real wall-clock win. No regressions found by any of the seven agents; the `[Medium]` items are deliberate, defensible trade-offs worth acknowledging, not defects.

- **Blockers:** None.
- **Non-blocking but worth addressing:** see Top Recommendations — primarily the compute-cost/critical-path trade-offs from Performance and the stale `@v4` version reference in two YAML comments.

---

## 1. Backward Compatibility -- SAFE ✅

No public API surface; coverage, baseline semantics, and artifact consumers are all preserved.

- **[Low]** Job names changed (`Runtime Tests` → `Runtime Tests (shard N)`, same for accessibility). Required status checks match by rendered check-run name, so when these jobs are eventually promoted to required (per the stack's "promote once stable" plan), the rules must reference the sharded names. Not a regression today — both jobs are non-blocking, no in-repo ruleset references the old names, and `automerge.needs` doesn't include them.
- **[Low]** `fail-fast: false` means one bad example now surfaces as up to 2/4 failing checks instead of one — behaviorally correct, just a UI expectation change.
- Verified correct: each shard loads the full `kompendium.json`, so the union of `--shard=k/N` partitions covers all ~440 examples and the `> 400` enumeration guard still runs; the baseline writer is gated on `UPDATE_AXE_BASELINE` (unset in CI), so sharding can't corrupt it; the new artifact name `example-docs` doesn't collide with the `docs` job's `pr-docs`.
- **Positives:** Coverage and baseline semantics are provably preserved across the new shard boundaries.

## 2. Code Quality -- GOOD ✅

YAML is correct end-to-end; the one real issue is a stale version number inside a comment.

- **[Low]** The comments at `pr-checks.yml:114-115` and `:157-158` say "upload-artifact@**v4** rejects duplicate names within a run" while the step pins `# v7.0.1`. The behavioral claim is true (the constraint landed in v4 and persists), but the version reference is drift — say "upload-artifact" version-agnostically or match the pin. The PR body's "pinned to `@v4` tags" note has the same staleness and is worth editing.
- **[Low]** Three near-identical job preambles (checkout / set-up-node / `npm ci` / Playwright cache+install) now exist — but the duplication pattern pre-existed on the target branch; this PR didn't increase the copy count, only the third copy's divergence. (Architecture flags the same thing from the extensibility angle.)
- Verified correct: shard denominators match the matrices exactly (`[1,2]`↔`/2`, `[1,2,3,4]`↔`/4`); `npm run … -- --shard=k/N` appends cleanly after the positional `runtime`/`accessibility` filter; `docs:build` → `www/` → upload → download → `serve-www.cjs` path chain is consistent.
- **Positives:** Shard/matrix wiring is exactly correct, and the PR opportunistically tightens action pinning to SHAs with version comments.

## 3. Architecture -- GOOD ✅

Build-once/fan-out topology is sound: deterministic shard partitioning (sorted `TAGS` identical everywhere), and a `needs:` graph that fails closed — if the build job breaks, shards are skipped, not falsely green.

- **[Low]** Shard count is duplicated per suite (matrix list at `pr-checks.yml:90-93`/`:133-136` vs. `--shard=…/N` denominator at `:112`/`:155`). Changing one without the other silently breaks partitioning, and the `> 400` count guard won't catch it (it asserts enumeration, not per-shard execution). Inherent to the Actions-matrix + Playwright-shard pattern; a maintenance footgun, not a defect.
- **[Low]** "Re-run failed jobs" relies on the `example-docs` artifact still existing — fine within the 1-day retention, fails at the download step after it expires. Deliberate cost trade-off; a one-line comment would do.
- **[Low]** A future "visual" suite would copy the ~15-line job preamble a fourth time; a composite "prepare example-test runner" action would pay off then, not now.
- **Positives:** Fail-closed `needs:` fan-out, deterministic partitioning, and least-privilege permissions on every new job.

## 4. Security -- SAFE ✅

Strictly better than the target branch: pins the previously mutable `actions/cache@v5` tag to a SHA and adds minimal `contents: read` permissions to all three new/changed jobs.

- **[Low]** Permissions parity: pre-existing `lint`/`build`/`test` jobs still have no `permissions:` block and there's no workflow-level default — a pre-existing gap this PR doesn't touch. Optional future hardening: top-level `permissions: contents: read`.
- **[Low]** The SHA pins are internally consistent (same SHA ↔ same version comment across all workflow files), but they weren't verified against upstream — worth one eyeball against the official `actions/*` release tags.
- No issues in the high-risk areas: trigger is `pull_request` (not `pull_request_target`), no `secrets.*` references in new jobs, artifact share is same-run name-scoped (no poisoning vector or new trust boundary), `matrix.shard` is author-controlled static integers (no injection), and the artifact is public-source docs (no sensitive exposure).
- **Positives:** Net security hardening — SHA-pinning a mutable tag plus least-privilege permissions — with no new exposure path.

## 5. Observability -- GOOD ✅

Failure signal is well-preserved: per-shard job names, per-shard trace artifacts, `fail-fast: false`, and a docs-build failure now surfaces as its own clearly named red check.

- **[Low]** Stale `@v4` version in the trace-upload comments — same finding as Code Quality; the per-shard artifact naming itself is correctly *necessary* (without it, two failing shards would collide and the second upload would error, losing traces).
- **[Low]** No HTML report artifact — diagnosability rests on the `github` reporter's inline annotations plus trace zips (`npx playwright show-trace`). Pre-existing choice, not regressed.
- **[Low]** Traces capture DOM/console/network, but the suite only hits the locally served static docs with no auth or secrets in the env — nothing sensitive expected; unchanged by this PR.
- **Positives:** Per-shard naming keeps every shard's failure independently attributable, and `fail-fast: false` preserves signal from all shards.

## 6. Performance -- GOOD ✅

The wall-clock win is real, but it buys it with more total compute and a serialized prefix — deliberate trade-offs the PR description doesn't fully spell out.

- **[Medium]** Total runner-minutes increase substantially: per-job fixed setup (checkout + `npm ci` + cache restore, ~1 min) is now paid 7× (1 build + 6 shards) instead of 2× — roughly +5 runner-minutes of pure setup per PR run, plus artifact transfer. Defensible (wall-clock matters more for PR feedback) but should be acknowledged.
- **[Medium]** The critical path is serialized: every test job `needs:` the ~4-min build job (setup + ~3-min `docs:build`) before starting its own ~1-min setup. Previously the build overlapped with each job's own setup. For accessibility the net win is still clear (~9 min → ~7.5 min estimated); for the runtime suite (~5m19s before, ~4-min serial prefix alone) the gain may be marginal — worth confirming against a real run.
- **[Medium]** Playwright `--shard` splits by test *count* over the alphabetically sorted tag list, not by duration. Heavy examples (tables, editors, composite playgrounds — the a11y spec waits up to 5s for `networkidle` per example) clustering alphabetically can leave one shard much slower than its siblings, capping the effective speedup below the ideal N×. Inherent to `--shard`; data-dependent.
- **[Low]** `www/` artifact is 1 upload + 6 downloads of the full docs site — unquantified but compressed, 1-day retention, and almost certainly cheaper than the 3-min rebuild it replaces.
- **[Low]** `npm ci` per test job is unavoidable (only `www/` is shared); npm and Playwright-browser caches mitigate it.
- **Positives:** Splitting across jobs is the correct lever given `workers: '50%'` ≈ 1 worker on 2-core runners — within-job parallelism genuinely can't be increased.

## 7. Lime Platform Issues -- SAFE ✅

The lime-issue-detector catalog exclusively targets Lime CRM platform *runtime* hazards (UoW hooks, event cascades, endpoint/task threadpool blocking, Stencil runtime, DB access); a CI-workflow YAML change touches none of those execution contexts.

- No findings — no LPID rule applies to this diff.
- **Positives:** SHA-pinned actions with version comments are good supply-chain hygiene, albeit outside this dimension's rule scope.

---

## Overall Verdicts

| Dimension | Verdict |
|---|---|
| Backward Compatibility | **SAFE** ✅ |
| Code Quality | **GOOD** ✅ |
| Architecture | **GOOD** ✅ |
| Security | **SAFE** ✅ |
| Observability | **GOOD** ✅ |
| Performance | **GOOD** ✅ |
| Lime Platform Issues | **SAFE** ✅ |
| **Merge Readiness** | **READY TO MERGE** ✅ |

## Top Recommendations

1. **[Medium from Performance]** Acknowledge the compute-cost trade in the PR description (or a comment): setup overhead is now paid ~7× instead of 2× — roughly +5 runner-minutes per run bought for wall-clock. A deliberate trade, but it should be on the record.
2. **[Medium from Performance]** Verify on a real CI run that the serialized ~4-min build prefix doesn't erode most of the sharding gain for the **runtime** suite specifically (it was only ~5m19s total before; the a11y win is clear, the runtime win may be marginal at ×2).
3. **[Medium from Performance]** Watch shard balance on real runs — `--shard` splits by count over the alphabetical tag list, so adjacent heavy examples can pin one shard. If one shard consistently dominates, consider reordering or a duration-aware split later.
4. **[Low from Code Quality]** Fix the two YAML comments saying "upload-artifact@v4 rejects duplicate names" (`pr-checks.yml:114-115`, `:157-158`) to be version-agnostic or match the pinned `v7.0.1`, and update the PR body's stale "pinned to `@v4` tags" note.
5. **[Low from Architecture]** Add a one-line comment noting that re-running a failed shard after the 1-day artifact retention will fail at the download step (rebuild requires a full re-run).
6. **[Low from Backward Compatibility]** When promoting these jobs to required checks, remember the check names are now per-shard (`Runtime Tests (shard 1)`, `… (shard 2)`, and four accessibility entries).
7. **[Low from Security]** One-time eyeball of the four pinned SHAs against the official `actions/*` release tags (verified internally consistent across the repo's workflows, not verified against upstream).

🤖 Generated with [Claude Code](https://claude.com/claude-code) — 6-agent-review skill (7 agents incl. lime-issue-detector)

**Comment by adrianschmidt** [comment]
@coderabbitai review

File: .github/workflows/pr-checks.yml [hunk 1]
```diff
@@ -56,75 +56,111 @@ jobs:
...
+    - uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c # v8.0.1
+      with:
+        name: example-docs
+        path: www
+    - run: npm run test:examples:runtime -- --shard=${{ matrix.shard }}/2
     # Retained on failure (`trace: 'retain-on-failure'`); upload so the trace
-    # viewer (DOM snapshots, console, network) is reachable from the run.
+    # viewer (DOM snapshots, console, network) is reachable from the run. Named
+    # per shard because upload-artifact@v4 rejects duplicate names within a run.
     - name: Upload Playwright traces
```
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply .github/workflows/pr-checks.yml]: @coderabbitai fixed in ee60467ef — dropped the version number rather than updating it. The comment is worth keeping (it explains why the trace artifact is named per shard, so nobody "simplifies" it back to a static name and breaks parallel uploads), but the `@v4` was both wrong (pinned to v7.0.1) and prone to rot — a version in a comment won't be bumped by Dependabot and is easy to miss. The duplicate-name rejection holds for upload-artifact v4+ regardless, so the version added nothing. Now reads: "Named per shard because upload-artifact rejects duplicate names within a run."

[the remark this answers is not in the document]
Comment (adrianschmidt) [reply .github/workflows/pr-checks.yml]: Thanks for confirming, @coderabbitai — and for the original catch. Agreed on the durability point; that's exactly why we dropped the version rather than bumping it. Resolved.

**Comment by adrianschmidt** [comment]
## Review follow-up

READY TO MERGE with no blockers — the `[Medium]`s are deliberate trade-offs. Addressed the documentation/verification items; backing the rest with real CI data.

### ✅ Addressed
- **Rec 4 (stale `@v4`).** The two YAML trace-upload comments were already fixed in `ee60467ef` (version dropped → "…upload-artifact rejects duplicate names…"). The PR body's stale "pinned to `@v4` tags / `cache@v5` style" bullet is now corrected — everything is SHA-pinned (cache v5.0.5, upload v7.0.1, download v8.0.1).
- **Rec 1 (compute-cost trade).** Now on the record in the PR body: wall-clock improves but per-job setup is paid ~7× instead of 2× — deliberate, PR-feedback latency over runner-minutes.

### 📊 Verified with real CI data
- **Rec 2 (is the runtime win marginal?).** No — measured on CI: build ~1.2 min + slowest shard ~2.8 min ≈ **~4 min critical path**, vs ~5m19s before. Modest but real — and build-once is independently justified by accessibility (~9 min → ~4 min). Keeping runtime at ×2 also stops it becoming the long pole next to the four a11y shards.
- **Rec 3 (shard balance).** Balanced in practice: runtime shards 2.6–2.8 min, accessibility 2.3–2.9 min — no shard dominates. Count-based `--shard` is holding up; I'll only reach for a duration-aware split if a real imbalance shows up.
- **Rec 7 (SHA provenance).** Checked all three pins against upstream release tags: `cache@v5.0.5` = `27d5ce7`, `upload-artifact@v7.0.1` = `043fb46`, `download-artifact@v8.0.1` = `3e5f45b` — all match.

### 👍 Acknowledged (no change)
- **Rec 5 (1-day retention vs re-runs).** Real but self-announcing — a stale re-run fails loudly at the download step. Leaving the cheap 1-day default rather than adding an edge-case comment; happy to add a one-liner if preferred.
- **Rec 6 (per-shard check names when promoting to required).** Noted — when these graduate from non-blocking, the branch rule must reference `Runtime Tests (shard 1/2)` and the four `Accessibility Tests (shard N)` entries, not the old names.

**Comment by adrianschmidt** [comment]
Correction to my earlier note above — I oversold the cost/benefit framing.

The ~7× setup increase is the cost of the **whole** optimization, and it buys the **combined** ~10 min of wall-clock reduction across both suites — overwhelmingly from **accessibility** (~9m06s → ~4 min) plus building the docs **once** instead of in both jobs. The **runtime** split is the marginal piece: ~5m14s → ~4m05s, i.e. ~2× setup for ~1 min, kept mainly so runtime doesn't become the long pole next to the four accessibility shards.

Pairing the full 7× against the runtime minute (as I implied) was wrong — that minute is bought by ~2× setup, not 7×. PR body updated with the honest accounting.

**Review by Kiarokh — APPROVED** [review]
